atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweepThread.java

```diff
@@ -0,0 +1,326 @@
+        } catch (InsufficientConsistencyException e) {
+            log.warn("Could not sweep because not all nodes of the database are online.", e);
+            return SweepOutcome.NOT_ENOUGH_DB_NODES_ONLINE;
+        } catch (RuntimeException e) {
+            specificTableSweeper.updateSweepErrorMetric();
+
+            return determineCauseOfFailure(e, tableToSweep.get());
+        }
+    }
+
+    private Optional<TableToSweep> getTableToSweep() {
+        return specificTableSweeper.getTxManager().runTaskWithRetry(
+                tx -> {
+                    Optional<SweepProgress> progress = currentTable.flatMap(
+                            tableToSweep -> specificTableSweeper.getSweepProgressStore().loadProgress(
+                                    tableToSweep.getTableRef()));
+                    SweepPriorityOverrideConfig overrideConfig = sweepPriorityOverrideConfig.get();
+                    if (progress.map(
+                            realProgress -> shouldContinueSweepingCurrentTable(realProgress, overrideConfig))
+                            .orElse(false)) {
+                        try {
+                            createLockForTableIfNecessary(progress.get());
+                            if (currentTable.isPresent()) {
+                                //noinspection ConstantConditions // class runs in a single thread, so this is fine
+                                currentTable.get().refreshLock();
+                            }
+                            return currentTable;
+                        } catch (InterruptedException ex) {
+                            log.info("Sweep lost the lock for table {}",
+                                    LoggingArgs.tableRef(progress.get().tableRef()));
+                            currentTable = Optional.empty();
+                            // We'll fall through and choose a new table
+                        }
+                    }
+
+                    log.info("Sweep is choosing a new table to sweep.");
+                    return getNextTableToSweep(tx, overrideConfig);
+                });
+    }
+
+    // This is needed if we've just grabbed the sweep thread lock and a previous thread was in the middle of a table
+    private void createLockForTableIfNecessary(SweepProgress progress) {
+        if (currentTable.isPresent() && !currentTable.get().getTableRef().equals(progress.tableRef())) {
+            // If currentTableToSweep is present, we're stopping the current table due to a new override. Release the lock
+            currentTable.ifPresent(tableToSweep -> tableToSweep.getSweepLock().close());
+        }
+
+        TableReference tableRef = progress.tableRef();
+        SingleLockService singleLockService = SingleLockService.createNamedLockServiceForTable(
+                lockService, TABLE_LOCK_PREFIX, tableRef);
+        currentTable = Optional.of(TableToSweep.continueSweeping(tableRef, singleLockService, progress));
```

Comment:
There was a bug in the "keeping track of current table logic" (see the last commit), partly because the logic is split over a few classes.
* `TableToSweep.continueSweeping` sets the `hasPreviousProgress` flag to true. Even though we're sweeping the same table, we need to update the progress, which at present means creating a whole new TableToSweep object (and creating a new `SingleLockService` too). If you don't do this, then the issue shows up in [a completely different part of the code](https://github.com/palantir/atlasdb/pull/3257/files#diff-94bc098d7947536a70c47437788b4849R219) (`SpecificTableSweeper.processFinishedSweep`), causing the sweep.priority table to be continually updated. Either this or another side-effect caused a sweep of the sweep.priority table to loop infinitely.

I'm considering a refactor to pass some of these responsibility into `TableToSweep`, or perhaps a new class `SweepState` which manages the sweep state machine and holds the `Optional<TableToSweep>` object. This class would be responsible for opening, refreshing and closing the per-table locks, as well as keeping the sweep progress state updated. However, this PR is large already, so perhaps this is best deferred to a later date.